cmds/watch.c

```diff
@@ -27,25 +27,43 @@ bool w_cmd_realpath_root(json_t *args, char **errmsg)
   return true;
 }
 
-/* clock /root
+/* clock /root [options]
  * Returns the current clock value for a watched root
+ * If the options contain a sync_timeout, we ensure that the repo
+ * is synced up-to-date and the returned clock represents the
+ * latest state.
  */
 static void cmd_clock(struct watchman_client *client, json_t *args)
 {
   w_root_t *root;
   json_t *resp;
-
-  /* resolve the root */
-  if (json_array_size(args) != 2) {
+  unsigned int sync_timeout = 0;
+
+  if (json_array_size(args) == 3) {
+    json_t *options = json_array_get(args, 2);
+    if (options) {
+      json_t *st_json = json_object_get(options, "sync_timeout");
+      if (st_json) {
+        sync_timeout = (unsigned int)json_integer_value(st_json);
+      }
+    }
+  }
+  else if (json_array_size(args) != 2) {
     send_error_response(client, "wrong number of arguments to 'clock'");
     return;
   }
 
+  /* resolve the root */
   root = resolve_root_or_err(client, args, 1, false);
   if (!root) {
     return;
   }
 
+  if (sync_timeout) {
+    w_root_sync_to_now(root, sync_timeout);
+    root->ticks++;
```

Comment:
From my testing, w_root_sync_to_now pre-increments ticks, such that when the cookie event arrives, if it has many other changes inside it, then that tick will be wrapped up with those changes.  So if we modify 10,000 files and then sync, it's possible we'd get the clock that contains those 10,000 changes because clock ticks seem to be inclusive.

However I could be totally wrong about how this works, that's just how it seemed to me.